### PR TITLE
Hitting the linter loop limit should be treated as an error

### LINKED ISSUE #1386: When linter loop limit is hit, should report an error/exception
When the loop limit is hit, `sqlfluff fix` logs a warning:
```
Loop limit on fixes reached
```

This is a pretty "quiet" notification that can mask issues. Suggest we change this:
* Increase the default loop limit from 10 to 25 iterations.
* Log an ERROR and/or raise an exception. If an exception, suggest (maybe?) raising one that is caught and included in the list of lint issues, not one that stops the `fix` run. Open to other ideas, but I think this is the right tradeoff between raising visibility of what is _likely_ an infinite loop in the linter run while still giving the user useful `fix` output.

Related to #1344, among others.

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #1386

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/core/linter/linter.py [hunk 1]
```diff
@@ -542,8 +543,27 @@ def lint_fix_parsed(
                     "loops."
                 )
                 break
-        if fix and loop + 1 == loop_limit:
-            linter_logger.warning(f"Loop limit on fixes reached [{loop_limit}].")
+        else:
+            if fix:
```
Comment (barrywhart) [inline src/sqlfluff/core/linter/linter.py:547]: The old `if` is considered a poor, hacky way to determine if the loop exited or not. I updated it to use `for/else` instead. Explanation: https://book.pythontips.com/en/latest/for_-_else.html#else-clause

File: src/sqlfluff/core/rules/base.py [hunk 2]
```diff
@@ -555,39 +555,28 @@ def crawl(
...
-            if lerr:
-                if ignore_mask:
-                    filtered = LintedFile.ignore_masked_violations([lerr], ignore_mask)
-                    if not filtered:
-                        lerr = None
-                        ignored = True
-            if lerr:
-                new_lerrs.append(lerr)
-            if not ignored:
-                new_fixes.extend(res.fixes)
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/base.py:574]: I moved the nested function `_process_lint_result()` to a separate method on `RuleBase` so the automated test can use it.

File: src/sqlfluff/core/linter/linter.py [hunk 3]
```diff
@@ -483,7 +483,8 @@ def lint_fix_parsed(
         else:
             ignore_buff = []
 
-        for loop in range(loop_limit):
+        save_tree = tree
+        for loop in range(loop_limit + 1):
```
Comment (barrywhart) [inline src/sqlfluff/core/linter/linter.py:487]: When I changed to a `for/else` loop (see the comment below for more info), I had to adjust the loop range in order to keep the same behavior -- one of the existing tests surfaced this subtle issue.

File: test/fixtures/rules/std_rule_cases/L019.yml [hunk 4]
```diff
@@ -230,10 +230,10 @@ trailing_comma_move_past_several_comment_lines:
     global_actions_states
   configs:
     core:
-      # Set runaway_limit=1 to verify the fix only requires one pass. In an
+      # Set runaway_limit=2 to verify the fix only requires one pass. In an
       # earlier version, the comma before "SAFE_DIVIDE()" was being moved one
       # line per pass. Too lazy!
-      runaway_limit: 1
+      runaway_limit: 2
```
Comment (barrywhart) [inline test/fixtures/rules/std_rule_cases/L019.yml:236]: Now that we treat hitting the limit as an error, these tests need one additional loop (First iteration generates a fix, then the second iteration generates **no** fixes and the loop exits). One iteration would've not only caused the linter to report an error, but also to discard the fixes!)

**Comment by barrywhart** [comment]
@juhoautio: Given there is work remaining to do on #2624 (i.e. rules that are currently causing the additional lint check to fail), I think your suggestion about the linter loop should be a separate issue after all. It's a great idea, but we shouldn't add it while there are still known issues.

File: src/sqlfluff/core/linter/linter.py [hunk 5]
```diff
@@ -542,8 +543,31 @@ def lint_fix_parsed(
...
+                # because it means their file(s) weren't fixed.
+                for violation in initial_linting_errors:
+                    if isinstance(violation, SQLLintError):
+                        violation.fixes = []
+
+                # Return the original parse tree, before any fixes were applied.
+                # Reason: When the linter hits the loop limit, the file is often
+                # messy, e.g. some of the fixes were applied repeatedly, possibly
+                # other weird things. We don't want the user to see this junk!
+                return save_tree, initial_linting_errors, ignore_buff
```
Comment (alanmcruickshank) [inline src/sqlfluff/core/linter/linter.py:570]: I like this approach 👍

**Review by alanmcruickshank — APPROVED** [review]
This looks good. Thanks @barrywhart

**Pull request merged**